apis/experimental/node_classes/types.go

```diff
@@ -0,0 +1,152 @@
+package node_classes
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/api"
+)
+
+// This should match kubeadm's API as closely as possible, so its definition is
+// uninteresting and not fleshed out.
+//
+// What is more important (for this PR) is the representation of nodes and how
+// they are linked to the concept of the Cluster.
+type Cluster struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+
+	// ...
+}
+
+// This is the level that cloud-agnostic controllers would interact with. This
+// is where one could set the desired size/shape of a NodePool.
```

Comment:
I think this should provide more definition. In particular, a `NodePool` is a collection of homogeneous nodes that are all part of a Kubernetes cluster - it's something that should be deployable using a single IG/ASG.

Reply:
That is a good requirement to state outright.

One of the other things we had discussed (but I haven't heard a strong desire for yet) was the possibility to have the semantics of a NodeTemplate instead of a NodePool. Any time you wanted a new node, it would copy all of the relevant attributes from the corresponding template, but you could update the template without necessarily updating all of the nodes that were created from it. After updating, new nodes created from the template would have the new attributes. That might make it easier to create one-off node types for experimentation without the overhead of always having to create a new NodePool of size 1, although we could also support orphaned nodes that weren't part of any group.